### PR TITLE
Make pathology tumor detection bundle deterministic and using internal pre-trained ResNet18

### PR DESCRIPTION
**Pull request by bhashemian**
### Description
This PR makes the pathology tumor detection bundle deterministic but setting the deabling `torch.backends.cudnn.benchmark` flag. It also train the model using an internal pretrained weight for ResNet18.

### Status
**Ready**

### Please ensure all the checkboxes:

- [x] Codeformat tests passed locally by running `./runtests.sh --codeformat`.
- [x] In-line docstrings updated.
- [x] Update `version` and `changelog` in `metadata.json` if changing an existing bundle.
- [x] Please ensure the naming rules in config files meet our requirements (please refer to: `CONTRIBUTING.md`).
- [x] Ensure versions of packages such as `monai`, `pytorch` and `numpy` are correct in `metadata.json`.
- [x] Descriptions should be consistent with the content, such as `eval_metrics` of the provided weights and TorchScript modules.
- [x] Files larger than 25MB are excluded and replaced by providing download links in `large_file.yml`.
- [x] Avoid using path that contains personal information within config files (such as use `/home/your_name/` for `"bundle_root"`).

### CODE REVIEW

File: models/pathology_tumor_detection/configs/multi_gpu_train.json [hunk 1]
```diff
@@ -29,7 +29,6 @@
         "$dist.is_initialized() or dist.init_process_group(backend='nccl')",
         "$torch.cuda.set_device(@device)",
         "$monai.utils.set_determinism(seed=123)",
```
Comment (yiheng-wang-nv) [inline models/pathology_tumor_detection/configs/multi_gpu_train.json]: `"$monai.utils.set_determinism(seed=123)",`
should be
`"$monai.utils.set_determinism(seed=123)"`

Comment (Nic-Ma) [reply models/pathology_tumor_detection/configs/multi_gpu_train.json]: @drbeh , as we are still building the CICD of execution tests in this repo, please ensure to run the configs.

Thanks.

File: models/pathology_tumor_detection/configs/train.json [hunk 2]
```diff
@@ -372,8 +372,7 @@
         }
     },
     "initialize": [
-        "$monai.utils.set_determinism(seed=15)",
-        "$setattr(torch.backends.cudnn, 'benchmark', True)"
+        "$monai.utils.set_determinism(seed=123)"
```
Comment (yiheng-wang-nv) [inline models/pathology_tumor_detection/configs/train.json]: Hi @drbeh , could you also update the weights and figures trained from deterministic settings?

Comment (bhashemian) [reply models/pathology_tumor_detection/configs/train.json]: Sure, but it will take a while as we are updating the pre-trained weights to nv-imagenet too.

**Comment by bhashemian** [comment]
updating with nv-image-net.

File: models/pathology_tumor_detection/configs/metadata.json [hunk 3]
```diff
@@ -2,6 +2,7 @@
     "schema": "https://github.com/Project-MONAI/MONAI-extra-test-data/releases/download/0.8.1/meta_schema_20220324.json",
     "version": "0.5.1",
```
Comment (yiheng-wang-nv) [inline models/pathology_tumor_detection/configs/metadata.json]: plz also update this number

**Comment by yiheng-wang-nv** [comment]
Hi @drbeh , please also add the descriptions of how to use pretrained weights from a local source in readme, and update the train config to address that.
You can also refer to:
https://github.com/Project-MONAI/model-zoo/tree/dev/models/endoscopic_tool_segmentation#pre-trained-weights
https://github.com/Project-MONAI/model-zoo/tree/dev/models/pathology_nuclei_segmentation_classification#model-overview

File: models/pathology_tumor_detection/configs/train.json (hunk 2 above)
Comment (bhashemian) [reply models/pathology_tumor_detection/configs/train.json]: I have updated the weights and figure for the deterministic training. We have postponed training with local weights, so maybe we can merge this till we have update with the nv-imagenet. Thanks

**Comment by bhashemian** [comment]
Hey @yiheng-wang-nv, this PR is ready to merge and the weights and figures are updated. It is still uses imagenet pertained weights but the training is deterministic. We don't have retraining with nv-imagenet yet so will update once we have it. Thanks

**Comment by Nic-Ma** [comment]
Hi @drbeh ，

Is the training in progress? maybe we can wait until we finalized the bundle?
If not, I just want to understand the blocker.

Thanks.

**Comment by bhashemian** [comment]
> Hi @drbeh ，
> 
> Is the training in progress? maybe we can wait until we finalized the bundle? If not, I just want to understand the blocker.
> 
> Thanks.

Hey @Nic-Ma, the GPU nodes have been busy so currently we are not training anything. Thanks

**Comment by bhashemian** [comment]
@yiheng-wang-nv @Nic-Ma this PR is updated with our internal weights, and is ready for review. Thanks

**Review by Nic-Ma — APPROVED** [review]

File: models/pathology_tumor_detection/docs/README.md [hunk 4]
```diff
@@ -39,6 +39,19 @@ The training was performed with the following:
 - Loss: BCEWithLogitsLoss
 - Whole slide image reader: cuCIM (if running on Windows or Mac, please install `OpenSlide` on your system and change `wsi_reader` to "OpenSlide")
 
+By setting the `"pretrained"` parameter of `TorchVisionFCModel` in the config file to `true`, ImageNet pre-trained weights will be used for training. Please note that these weights are for non-commercial use. Each user is responsible for checking the content of the models/datasets and the applicable licenses and determining if suitable for the intended use. In order to use other pretrained weights, you can use `CheckpointLoader` in train handlers section as the first handler:
```
Comment (yiheng-wang-nv) [inline models/pathology_tumor_detection/docs/README.md:44]: Hi @zephyrie , could you please also help to review this part? Thanks!

Comment (zephyrie) [reply models/pathology_tumor_detection/docs/README.md:44]: This looks good! I might suggest a header of "### Pretained Weights" so it's easier to see.  Also, I think it would make sense to move up the "Notes on Determinism" section to be under training configuration rather than MONAI Bundle Commands. You could just add it to the end of the "Training Configuration" section, that way all of the training caveats are in a single place.

File: models/pathology_tumor_detection/docs/README.md [hunk 5]
```diff
@@ -39,6 +39,19 @@ The training was performed with the following:
 - Loss: BCEWithLogitsLoss
 - Whole slide image reader: cuCIM (if running on Windows or Mac, please install `OpenSlide` on your system and change `wsi_reader` to "OpenSlide")
```
Comment (zephyrie) [inline models/pathology_tumor_detection/docs/README.md:41]: ### Pretained Weights

Comment (zephyrie) [inline models/pathology_tumor_detection/docs/README.md:41]: ### Pretained Weights

File: models/pathology_tumor_detection/docs/README.md [hunk 6]
```diff
@@ -54,9 +67,9 @@ Inference is performed on WSI in a sliding window manner with specified stride.
 ## Performance
```
Comment (zephyrie) [inline models/pathology_tumor_detection/docs/README.md:73]: You could move the determinism piece above this line: 

"### Note on determinism

By default this bundle use a deterministic approach to make the results reproducible. However, it comes at a cost of performance loss. Thus if you do not care about reproducibility, you can have a performance gain by replacing `"$monai.utils.set_determinism"` line with `"$setattr(torch.backends.cudnn, 'benchmark', True)"` in initialize section of training configuration (`configs/train.json` and `configs/multi_gpu_train.json` for single GPU and multi-GPU training respectively)."

File: models/pathology_tumor_detection/docs/README.md (hunk 5 above)
Comment (bhashemian) [reply models/pathology_tumor_detection/docs/README.md:41]: Done!

File: models/pathology_tumor_detection/docs/README.md (hunk 6 above)
Comment (bhashemian) [reply models/pathology_tumor_detection/docs/README.md:73]: Done!

**Comment by yiheng-wang-nv** [comment]
HI @drbeh @Nic-Ma , I pushed two commits with small changes. The PR looks good to me, and now I'm checking the eval result, will approve and merge the PR after the check.

**Review by yiheng-wang-nv — APPROVED** [review]
LGTM, I also verified the weights

**Comment by KumoLiu** [comment]
Hi @drbeh, I don't think we could use the default value of this arg still be `True` here.
https://github.com/drbeh/model-zoo/blob/3f424fca53b4d80ee507db2928613cc827ffc4e9/models/pathology_tumor_detection/configs/train.json#L35

Hi @yiheng-wang-nv, and @Nic-Ma, could you please help double-confirm?
Thanks!

**Comment by yiheng-wang-nv** [comment]
> Hi @drbeh, I don't think we could use the default value of this arg still be `True` here. https://github.com/drbeh/model-zoo/blob/3f424fca53b4d80ee507db2928613cc827ffc4e9/models/pathology_tumor_detection/configs/train.json#L35
> 
> Hi @yiheng-wang-nv, and @Nic-Ma, could you please help double-confirm? Thanks!

Hi @KumoLiu , thanks for the finding. I think we need to set `False` here.

Hi @drbeh , may need your double confirm that the weights provided in `large_files.yml` are trained with internal pretrained ResNet18 rather than imagenet pretrained weights.

**Comment by bhashemian** [comment]
> > Hi @drbeh, I don't think we could use the default value of this arg still be `True` here. https://github.com/drbeh/model-zoo/blob/3f424fca53b4d80ee507db2928613cc827ffc4e9/models/pathology_tumor_detection/configs/train.json#L35
> > Hi @yiheng-wang-nv, and @Nic-Ma, could you please help double-confirm? Thanks!
> 
> Hi @KumoLiu , thanks for the finding. I think we need to set `False` here.
> 
> Hi @drbeh , may need your double confirm that the weights provided in `large_files.yml` are trained with internal pretrained ResNet18 rather than imagenet pretrained weights.

That is actually fine because the default for the pipeline is no using our internal pre-trained weights but using ImageNet. The provided weights in `large_files.yml` are trained with setting `pretrained=False` and I have already checked with both flags and they give the same results.
